File: crates/snipx-core/src/diagnostic.rs
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SourceSpan {
    pub start: usize,
    pub end: usize,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Severity {
    Error,
    Warning,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DiagnosticCode {
    ParseError,
    UnterminatedSnippet,
    UnterminatedString,
    UnterminatedBlockComment,
    UnterminatedIntralineaBlock,
    InvalidDirectivePosition,
    DuplicateDirective,
    InvalidLocalSubjectMarker,
    EmptyLocalSubject,
    InvalidCliUsage,
    MissingAmbientSubject,
    InvalidDecorationTarget,
    InvalidStatementTerminator,
    UnsupportedProfile,
    RawHtmlOmitted,
    InvalidSnippet,
    InvalidNumber,
    SnippetNotFound,
    SnippetAmbiguous,
    // Fragility lint warnings. Provisional pending ratification of
    // ADR 0004 (docs/adr/0004-fragility-diagnostics.md).
    FragileShortAnchor,
    FragileNearDuplicate,
    FragileCaptureContext,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RelatedSpan {
    pub message: String,
    pub span: SourceSpan,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Diagnostic {
    pub code: DiagnosticCode,
    pub severity: Severity,
    pub message: String,
    pub span: Option<SourceSpan>,
    pub related: Vec<RelatedSpan>,
}

File: packages/snipx-ts/src/diagnostic.ts
/** Mirror of crates/snipx-core/src/diagnostic.rs. */

export interface SourceSpan {
  start: number;
  end: number;
}

export type Severity = "error" | "warning";

export type DiagnosticCode =
  | "ParseError"
  | "UnterminatedSnippet"
  | "UnterminatedString"
  | "UnterminatedBlockComment"
  | "UnterminatedIntralineaBlock"
  | "InvalidDirectivePosition"
  | "DuplicateDirective"
  | "InvalidLocalSubjectMarker"
  | "EmptyLocalSubject"
  | "InvalidCliUsage"
  | "MissingAmbientSubject"
  | "InvalidDecorationTarget"
  | "InvalidStatementTerminator"
  | "UnsupportedProfile"
  | "RawHtmlOmitted"
  | "InvalidSnippet"
  | "InvalidNumber"
  | "SnippetNotFound"
  | "SnippetAmbiguous";

export interface RelatedSpan {
  message: string;
  span: SourceSpan;
}

export interface Diagnostic {
  code: DiagnosticCode;
  severity: Severity;
  message: string;
  span: SourceSpan | null;
  related: RelatedSpan[];
}

export const DIAGNOSTIC_CODE_NAMES: Record<DiagnosticCode, string> = {
  ParseError: "PARSE_ERROR",
  UnterminatedSnippet: "UNTERMINATED_SNIPPET",
  UnterminatedString: "UNTERMINATED_STRING",
  UnterminatedBlockComment: "UNTERMINATED_BLOCK_COMMENT",
  UnterminatedIntralineaBlock: "UNTERMINATED_INTRALINEA_BLOCK",
  InvalidDirectivePosition: "INVALID_DIRECTIVE_POSITION",
  DuplicateDirective: "DUPLICATE_DIRECTIVE",
  InvalidLocalSubjectMarker: "INVALID_LOCAL_SUBJECT_MARKER",
  EmptyLocalSubject: "EMPTY_LOCAL_SUBJECT",
  InvalidCliUsage: "INVALID_CLI_USAGE",
  MissingAmbientSubject: "MISSING_AMBIENT_SUBJECT",
  InvalidDecorationTarget: "INVALID_DECORATION_TARGET",
  InvalidStatementTerminator: "INVALID_STATEMENT_TERMINATOR",
  UnsupportedProfile: "UNSUPPORTED_PROFILE",
  RawHtmlOmitted: "RAW_HTML_OMITTED",
  InvalidSnippet: "INVALID_SNIPPET",
  InvalidNumber: "INVALID_NUMBER",
  SnippetNotFound: "SNIPPET_NOT_FOUND",
  SnippetAmbiguous: "SNIPPET_AMBIGUOUS",
};
